### PR TITLE
Fix `true_divide` with dtype argument

### LINKED ISSUE #2059: Unexpected result in cupy.divide with dtype argument
Version: current master 202e4e8902e593b8c8ba487f8371e4b6a42e4660
```
>>> cupy.show_config()
CuPy Version          : 6.0.0b2
CUDA Root             : /usr/local/cuda
CUDA Build Version    : 9020
CUDA Driver Version   : 9020
CUDA Runtime Version  : 9020
cuDNN Build Version   : 7005
cuDNN Version         : 7005
NCCL Build Version    : 2104
NCCL Runtime Version  : (unknown)
```

#### Code
```py
import cupy
import numpy

a = cupy.array(500., dtype=numpy.float64)

c = cupy.divide(1., a)
print(c.dtype, c)

c = cupy.divide(1., a, dtype=numpy.float64)
print(c.dtype, c)
```
#### Result
```
float64 0.002
float64 -0.08333333333333333
```

### PR DESCRIPTION
**Pull request by okuta**
Fix #2059

### CODE REVIEW

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by niboshi** [comment]
How does this fix work?

**Comment by okuta** [comment]
```
for op in ops[::-1]:  # the trick to avoid using small type
```
In most cases, the largest type is chosen from the reverse order.

**Comment by niboshi** [comment]
@okuta 
I think we talked offline about implementing this in a different way. Is that correct?

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by okuta** [comment]
jenkins, test this please.

File: cupy/core/_kernel.pyx [hunk 1]
```diff
@@ -735,6 +736,7 @@ cdef class ufunc:
         readonly Py_ssize_t nargs
         readonly object name
         readonly list _ops
+        readonly list _out_ops
```
Comment (niboshi) [inline cupy/core/_kernel.pyx]: Please add some comment.

**Comment by okuta** [comment]
I updated.

File: cupy/core/_kernel.pyx [hunk 2]
```diff
@@ -770,7 +770,8 @@ cdef _min_scalar_type(x):
     return dt
 
 
-cdef tuple _guess_routine(str name, dict cache, list ops, list in_args, dtype):
+cdef tuple _guess_routine(
+        str name, dict cache, list ops, list in_args, dtype, list out_ops):
```
Comment (niboshi) [inline cupy/core/_kernel.pyx]: This function needs more comments, but not necessarily in this PR.

File: cupy/core/_kernel.pyx [hunk 3]
```diff
@@ -814,7 +815,8 @@ cdef class ufunc:
         readonly Py_ssize_t nout
         readonly Py_ssize_t nargs
         readonly object name
-        readonly list _ops
+        readonly list _ops  # normal routines
+        readonly list _out_ops  # routines based on output dtype
```
Comment (niboshi) [inline cupy/core/_kernel.pyx]: Is this more appropriate?

```suggestion
        readonly list _out_ops  # routines based on explicitly given output dtype
```

File: tests/cupy_tests/math_tests/test_arithmetic.py [hunk 4]
```diff
@@ -239,6 +199,84 @@ def test_binary(self, xp):
...
+        'arg2': [numpy.array([-3, -2, -1, 1, 2, 3], dtype=d)
+                 for d in negative_types
+                 ] + [0, 0.0, 0j, 2, 2.0, 2j, -2, -2.0, -2j, True, False],
+        'name': ['divide', 'true_divide', 'subtract'],
+    })
+))
+class TestArithmeticBinary(ArithmeticBinaryBase, unittest.TestCase):
+
+    def test_binary(self):
+        self.use_dtype = False
```
Comment (niboshi) [inline tests/cupy_tests/math_tests/test_arithmetic.py]: This can be a class attribute.

File: tests/cupy_tests/math_tests/test_arithmetic.py [hunk 5]
```diff
@@ -239,6 +199,84 @@ def test_binary(self, xp):
...
+        self.check_binary()
+
+    @testing.with_requires('numpy>=1.10')
+    def test_binary_with_dtype(self):
+        self.use_dtype = True
+        self.check_binary()
+
```
Comment (niboshi) [inline tests/cupy_tests/math_tests/test_arithmetic.py]: How about parameterizing `use_dtype` (and `raise SkipTest()` conditionally).

Comment (okuta) [reply tests/cupy_tests/math_tests/test_arithmetic.py]: OK!

**Comment by okuta** [comment]
I fixed.
jenkins, test this please.

File: tests/cupy_tests/math_tests/test_arithmetic.py [hunk 6]
```diff
@@ -126,53 +126,10 @@ def test_imag(self, xp, dtype):
...
-                 for d in negative_no_complex_types
-                 ] + [0, 0.0, 2, 2.0, -2, -2.0, True, False],
-        'arg2': [numpy.array([-3, -2, -1, 1, 2, 3], dtype=d)
-                 for d in negative_no_complex_types
-                 ] + [0, 0.0, 2, 2.0, -2, -2.0, True, False],
-        'name': ['floor_divide', 'fmod', 'remainder'],
-    })
-))
-class TestArithmeticBinary(unittest.TestCase):
+class ArithmeticBinaryBase():
```
Comment (niboshi) [inline tests/cupy_tests/math_tests/test_arithmetic.py]: Nitpick 

```suggestion
class ArithmeticBinaryBase:
```

File: cupy/core/_kernel.pyx [hunk 7]
```diff
@@ -981,8 +986,8 @@ cdef class ufunc:
         return kern
 
 
-cpdef create_ufunc(name, ops, routine=None, preamble='', doc='',
-                   default_casting=None, loop_prep=''):
+cdef list _get_ops(ops, routine):
+    """Parse dtype character mapping"""
```
Comment (niboshi) [inline cupy/core/_kernel.pyx]: It's difficult to tell what's the difference between inputs and outputs. Can you improve it?

Comment (niboshi) [reply cupy/core/_kernel.pyx]: I think I'm going to address this later.

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
Can you resolve conflict?

**Comment by niboshi** [comment]
Jenkins, test this please

**Pull request merged**